tez-runtime-library/src/main/java/org/apache/tez/runtime/library/common/sort/impl/IFile.java

```diff
@@ -71,6 +73,223 @@
+      this.rawOut = newRawOut;
+
+      this.checksumOut = new IFileOutputStream(this.rawOut);
+      if (fileCodec != null) {
+        this.compressor = CodecPool.getCompressor(fileCodec);
+        if (this.compressor != null) {
+          this.compressor.reset();
+          this.compressOutput = true;
+          this.compressedOut = fileCodec.createOutputStream(checksumOut, compressor);
+          this.out = new FSDataOutputStream(this.compressedOut, null);
+        } else {
+          LOG.warn("Could not obtain compressor from CodecPool");
+          this.out = new FSDataOutputStream(checksumOut, null);
+        }
+      } else {
+        this.out = new FSDataOutputStream(checksumOut, null);
+      }
+
+      // Write header to file
+      headerWritten = false;
+      writeHeader(newRawOut);
+
+      // write real data
+      int sPos = HEADER.length;
+      int len = (bout.size() - checksumSize - HEADER.length);
+      this.out.write(bout.getBuffer(), sPos, len);
+
+      bufferFull = true;
+      bout.reset();
+    }
+
+    @Override
+    public void close() throws IOException {
+      try {
+        super.close();
+      } finally {
+        if (this.rawOut != null) {
+          this.rawOut.close();
+        }
+      }
+    }
+
+    boolean shouldWriteToDisk() {
+      return totalSize >= (cacheSize - (checksumSize +
+          (2 * WritableUtils.getVIntSize(EOF_MARKER))));
+    }
+
+    @Override
+    protected void writeKVPair(byte[] keyData, int keyPos, int keyLength,
+        byte[] valueData, int valPos, int valueLength) throws IOException {
+      if (!bufferFull) {
+        // Compute actual payload size: write RLE marker, length info and then entire data.
+        totalSize += ((prevKey == REPEAT_KEY) ? V_END_MARKER_SIZE : 0)
+            + WritableUtils.getVIntSize(keyLength) + keyLength
+            + WritableUtils.getVIntSize(valueLength) + valueLength;
+
+        if (shouldWriteToDisk()) {
+          resetToFileBasedWriter();
+        }
+      }
+      super.writeKVPair(keyData, keyPos, keyLength, valueData, valPos, valueLength);
+    }
+
+    @Override
+    protected void writeValue(byte[] data, int offset, int length) throws IOException {
+      if (!bufferFull) {
+        totalSize += ((prevKey != REPEAT_KEY) ? RLE_MARKER_SIZE : 0)
+            + WritableUtils.getVIntSize(length) + length;
+
+        if (shouldWriteToDisk()) {
+          resetToFileBasedWriter();
+        }
+      }
+      super.writeValue(data, offset, length);
+    }
+
+    /**
+     * Check if data was flushed to disk.
+     *
+     * @return whether data is flushed to disk ot not
+     */
+    public boolean isDataFlushedToDisk() {
+      return bufferFull;
+    }
+
+    /**
+     * Get cached data if any
+     *
+     * @return if data is not flushed to disk, it returns in-mem contents
+     */
+    public byte[] getData() {
+      if (!isDataFlushedToDisk()) {
+        byte[] buf = new byte[cacheStream.size()];
```

Comment:
Nit: Not sure if this really matters. Could avoid this copy by wrapping in a ByteBuffer (another copy is happening anyway when serializing to protobuf)

Reply:
Will fix this in next commit.